### DISCUSSION TITLE
Mode Solver:  saving mode data included fields

### DISCUSSION THREAD

**Discussion by smgitrep** [flexcompute/tidy3d#1657] [Q&A] [answered]
After running the mode solver in Tidy3D, the mode data (neff, ngroup, x, y, Ex, Ey, Ez, etc) seems to be returned as an xarray.  How can I save (and load) the xarray for post-processing in the future using Tidy3D functions and methods.

**Comment by tylerflex** [comment, marked as answer]
Hi @smgitrep the `ModeSolverData` can be written to file with `to_file` and read using `from_file`.

So for example, if
```py
mode_data = mode_solver.solve()

mode_data.to_file('mode_data.hdf5')

# and later

mode_data = td.ModeSolverData.from_file('mode_data.hdf5')
```

hope that helps

https://docs.flexcompute.com/projects/tidy3d/en/latest/api/_autosummary/tidy3d.components.base.Tidy3dBaseModel.html#tidy3d.components.base.Tidy3dBaseModel.to_file

**Reply by gaubry** [reply]
Hi @tylerflex ,
Is it possible to save and load the `ModeSolver` with the `ModeSolverData` included?
For instance to be able to use the `ModeSolver.plot_field()` or `ModeSolver.to_source()` methods without having to recompute the modes.
If I do ModeSolver.to_file(), the `ModeSolverData` are not saved and are recomputed when I load it.
Best,
Geoffroy

**Reply by momchil-flex** [reply]
I think the best is to switch to using `ModeSimulation`. This is the new way we'll be transitioning to running the mode solver, exactly in order to address such issues that are intrinsic in the original implementation. Unfortunately we haven't really documented this yet, because it is not supported in our GUI. But for python development, it is the most straightforward way to achieve what you want. There's an example usage e.g. in [this notebook](https://docs.flexcompute.com/projects/tidy3d/en/latest/notebooks/SiWaveguideTPA.html). 

The returned ModeSimulationData will store everything to file, and can be used to plot the fields.

**Reply by gaubry** [reply]
Thanks a lot. Just a question: the ModeSimulation is necessarily run on the server? Or it can be also locally run?

**Reply by momchil-flex** [reply]
It can be run locally with `.run_local()`.

**Reply by gaubry** [reply]
I'm still struggling... :-/
This is an example of what I'd like to do:

```
modeTaskName = "mode_coreRadius%.2fum"%(coreRadius)
modeFileName = "data/%s.hdf5"%modeTaskName

if os.path.isfile(modeFileName):
    ms = td.ModeSimulation.from_file(fname = modeFileName) # I'd like to load the data here
    # or:
    modes = td.ModeSimulationData.from_file(fname = modeFileName) # I'd like to load the data here
else:
    # position and size of source plane
    src_plane = td.Box(center=[0, 0, src_pos], size=[xy_plane, xy_plane, 0])

    # number of modes to compute
    num_modes = 4
    # setup and solve for modes
    mode_spec = td.ModeSpec(num_modes=num_modes)
    ms = td.ModeSimulation.from_simulation(
            sim_noNP,
            mode_spec=mode_spec,
            plane=src_plane,
            freqs = [freq0])
    modes = ms.run_local()
    # I'd like to save the data here
```
Problem : the cool functions to represent the modes and to transform it in a source exist for ModeSimulation. But if I save it the data are not saved, and are computed again.
Is my problem clearer?

**Reply by momchil-flex** [reply]
Save `modes.to_file()`. This contains both the data and the original mode simulation in `modes.simulation`. So you should be able to do e.g.

```
modes = td.ModeSimulationData.from_file(...)
modes.plot_field(...)
source = modes.simulation._mode_solver.to_source(...)
```

It actually seems like our `to_source` function is still only implemented for the `ModeSolver` object, which you can still access via the `_mode_solver` property. Granted, the last line is not the prettiest... But otherwise, the `ModeSimulationData` can be saved to file and should have everything you need?

**Reply by gaubry** [reply]
Perfect! Many thanks. For the sake of clarity, here the working example I now use thanks to your help:
```
modeTaskName = "mode_coreRadius%.2fum"%(coreRadius)
modeFileName = "data/%s.hdf5"%modeTaskName

if os.path.isfile(modeFileName):
    modes = td.ModeSimulationData.from_file(fname = modeFileName)
    num_modes = modes.simulation.mode_spec.num_modes
else:
    # position and size of source plane
    src_plane = td.Box(center=[0, 0, src_pos], size=[xy_plane, xy_plane, 0])

    # number of modes to compute
    num_modes = 4
    # setup and solve for modes
    mode_spec = td.ModeSpec(num_modes=num_modes)
    ms = td.ModeSimulation.from_simulation(
            sim_noNP,
            mode_spec=mode_spec,
            plane=src_plane,
            freqs = [freq0])
    modes = ms.run_local()
    modes.to_file(modeFileName)

modes.modes_raw.to_dataframe()
```

```
fig, axs = plt.subplots(num_modes, 3, figsize=(12, 10), tight_layout=True)
for mode_ind in range(num_modes):
    modes.plot_field("Ex", "real", f=freq0, mode_index=mode_ind, ax=axs[mode_ind, 0])
    modes.plot_field("Ey", "real", f=freq0, mode_index=mode_ind, ax=axs[mode_ind, 1])
    modes.plot_field("Ez", "real", f=freq0, mode_index=mode_ind, ax=axs[mode_ind, 2])
plt.show()
```

and I use the `modes.simulation._mode_solver.to_source()` method.

**Comment by smgitrep** [comment]
Thank you, Tyler!
